cmd/security-secrets-setup/main.go

```diff
@@ -16,80 +17,98 @@ package main
 
 import (
 	"flag"
-	"log"
+	"fmt"
 	"os"
 	"time"
 
 	"github.com/edgexfoundry/edgex-go/internal"
-	"github.com/edgexfoundry/edgex-go/internal/pkg/config"
 	"github.com/edgexfoundry/edgex-go/internal/pkg/usage"
 	"github.com/edgexfoundry/edgex-go/internal/security/setup"
-	"github.com/edgexfoundry/edgex-go/internal/security/setup/certificates"
-	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	"github.com/edgexfoundry/edgex-go/internal/security/setup/option"
 )
 
-func main() {
-	start := time.Now()
-	var configFile string
+type exit interface {
+	callExit(int)
+}
+
+type exitCode struct{}
```

Comment:
instead of defining a whole new type and interface for testing this kind of thing, what I've seen done in snapd is:

```go
var (
    ExitImpl = os.Exit
)

func main() {
    exitImpl(0)
}
```

then in tests you do this:

```go
func makeTestExit(exitedStatus *int) func(status int) {
    return func(status int){ *exitedStatus = status }
}

func TestTheThings() {
    exitCode := 0
    defer mockExit(makeTestExit(&exitCode))
    runMainWithErrorThings()
    if exitCode == 1 {
        // test pass
    } else {
        // test fail
    }
}

func mockExit(mockedExit func(status int)) func() {
    oldExit := pkg.ExitImpl
    pkg.ExitImpl = mockedExit
    return func() {
        pkg.ExitImpl = oldExit
    }
}
```

Reply:
First class functions!

Reply:
@bnevis-i  can you elaborate it more? What do you mean by "first class" functions?

Reply:
After some thinking, IMHO, I like interface approach better: the reasons are
1. Interface is clear to convey the idea what type of actions you are trying to do
2. Interface is also easier to put all collections of actions to do in one contract for the caller; though in this case we only have one but with that, it is easier to override the behavior in test and in other places if you find to do so.
3. IMHO, interface is more object-oriented than just using the variables

Moreover, like I said previously, the main purpose of setup and teardown is not to do the mocking but rather to initialize and clean up the things tests needed.